### ISSUE TITLE
dhall-any language binding generator?

### ISSUE THREAD

**Issue by SupraSummus** [dhall-lang/dhall-lang#295]
During reading of `semantics.md` and encoding the rules in dhall-python implementation I realized that I'm doing quite a mindless work there. I thought that a transpiler should be doing it instead of me. It'd be faster (once we have the transpiler), more reliable, repeatable and more fun (writing the transpiler). The transpiler can be generalised/parametrized to support many target languages.

However, the idea is very blurry at the moment and I don't know if it's doable right now.

### Parsing semantics specs

First step to generate bindings is to have a dhall specifications in machine readable format. I'm not sure, but to me it looks like we already have it. ABNF grammar is easy. Semantics specs are very formal and looks parsable. The only problem I see is that there is few example judgements that needs to be excluded. The fact that they are examples are encoded in natural language description near them.

I think I'll give it a try to write a parser for extracting judgements from plain `.md` files. It may be useful for other things apart transpiling. (For finding syntax errors ;) and ambiguities)

### Converting judgements into target language paradigm

As @Gabriel439 mentioned, one option here is to use [coq](https://coq.inria.fr/) code as intermediate step and generate bindings from it. Coq can extract code into OCaml, Haskell and Scheme however for python some patching is needed (http://coqcots.gforge.inria.fr/coq8.4pl3-python/).

~~Once we have the judgements loaded we need to convert them to form suitable for use in target language. For example:~~
 1. ~~for prolog there is nothing to change - judgements are isomorphic with prolog clauses~~
 2. ~~for imperative, non-objective languages transpiler needs to generate functions that through massive amount of `if`ing does type inference, shifting, substitutions and other.~~
 3. ~~for objective languages transpiler can take advantage of it by encoding expression types (not to confuse with `Type`s ;) ) as classes~~

~~This step looks reusable across languages, but not completely. For example there is many OOP semantics out there, so I don't know if transpiler will be inter-operable with all of them.~~

### ~~Encoding judgements in target language~~

~~Nothing interesting. Add boilerplate required for given target language. Braces, `def`s, `if`s, `var`s etc.~~

### Remaining work to be done on target language binding

Some aspects require language-specific or binding-specific operations:
 * creating and integrating the parser
 * integrating with cbor [de]serializer
 * wiring import routines

Let me know what you think. Maybe someone has already done it or at least described precisely what has to be done? Also pardon my limited-word-set English.

**Comment by Gabriel439** [comment]
@SupraSummus: I believe the closest thing to what you have in mind is "program extraction", meaning that the implementation is written in a programming language designed for program extraction (typically Coq) and then you can auto-generate code in other languages from that Coq code.

I've never done this myself, so somebody can correct me if I'm wrong, but the the idea is that if you require any high-performance operations (like machine arithmetic) you specify those in Coq as axioms and then you define all of the data types (i.e. `Expr`) and non-performance-critical functions (i.e. `β-normalize`) in Coq.  Then for program extraction it produces an intermediate language that you can compile into various languages.

So the first step would be converting the Dhall semantics to a Coq program and then the next step would be adding support for extracting Python programs from Coq.

**Comment by Gabriel439** [comment]
Actually, from some quick searching it looks like there might be a way to extract Python code from Coq: http://coqcots.gforge.inria.fr/coq8.4pl3-python/

**Comment by SupraSummus** [comment]
Thank you for pointers, @Gabriel439. So first I'll be trying to write `semantics.md` -> coq converter.

**Comment by SupraSummus** [comment]
Wait, coq looks great, but code generation in python is pain. I didn't tried it yet, but patching and building whole coq from source deters me.

Searching for "logic programming in python" I found pyDatalog and logpy. I wonder if they can be used for proving type and reduction queries. Not as a compilers/program extractors but inline target implementation.

**Comment by Profpatsch** [comment]
@SupraSummus Code generation has the disadvantage that you can’t directly influence how fast your implementation is going to be (especially when mapping from a functional paradigm to an imperative/interpreted paradigm).

A more fruitful approach should be: write one implementation in a fast language and then call that from other languages (especially if those are slow and interpreted, like python). The issue for that is https://github.com/dhall-lang/dhall-lang/issues/2

**Comment by SupraSummus** [comment]
another find: coq to llvm https://github.com/coq-ext-lib/coq-compile (looks a bit stale though)

@Profpatsch I belive that right optimizer can produce fast code. So more important than writing one fast implementation is to have an implementation that is optimizable (and easy do develop). That said I think that we have this implementation right here, in dhall-lang.

Writing implementation in C isn't a complete solution, because on some platforms you cannot run C code (ELF/DLL files, whatever). For example this is true in browser environment. (Of course you can transpile C to JS, but isn't it better to transpile Coq to JS?)

**Comment by Profpatsch** [comment]
> Of course you can transpile C to JS, but isn't it better to transpile Coq to JS?

In practice no, because there’s a lot more industry work going into mapping the C model to the browser world (WebAssembly) than into optimizing code generated from Coq.

**Comment by SupraSummus** [comment]
You may be right. I know coq for 3 hours and don't know how fast extracted code is nor how much work is going into developing coq.

**Comment by Gabriel439** [comment]
The main advantage of using program extraction over code generation is that the specification is type-checked (a.k.a. certified programming) and you don't have to reinvent a specification language for the semantics

**Comment by jmitchell** [comment]
I haven't used [Verified Software Toolchain (VST)](https://vst.cs.princeton.edu/) yet, but it claims to let you verify C programs by faithful translation to Coq via `clightgen` from the [CompCert](http://compcert.inria.fr/) project.

Workflow for using VST according chapter 3 of its [docs](https://vst.cs.princeton.edu/download/VC.pdf):

> * Write a C program `F.c`
> * Run `clightgen -normalize F.c` to translate it into a Coq file `F.v`.
> * Write a verification of `F.v` in a file such as `verif_F.v`. That latter file must import both `F.v` and the VST _Floyd_ program verification system, `VST.floyd.proofauto.

Then you can either use the C program as though the proof machinery doesn't exist, or you can compile to various architectures using the verified CompCert compiler. Before going with the latter choice by default, be sure to consider whether CompCert's license is agreeable.

**Comment by indolering** [comment]
I think what you are looking for is [WebAssembly Interface Types](https://hacks.mozilla.org/2019/08/webassembly-interface-types/), which (in theory) will handle type checking, type conversion, and calling conventions.  Wasmer has modules for [11 different languages](https://github.com/wasmerio/wasmer#languages) alone.  You might need to wrap file handling due to how different WASI operates, but the work to create a package boils down to this:

```
from wasmer import Instance

wasm_bytes = open('simple.wasm', 'rb').read()
instance = Instance(wasm_bytes)
exports = instance.exports
```

The proposal was only publicly announced 2 months ago, so you will need to wait a bit.

**Comment by Gabriel439** [comment]
@indolering: I believe we need something else in addition to WebAssembly interface types.  The interface types do support some declarative transformations, but we essentially need a way to encode arbitrary logic (e.g. Dhall's β-normalization algorithm, for example) in a machine-readable way so that we can generate β-normalization code in each target language.

**Comment by indolering** [comment]
> I believe we need something else in addition to WebAssembly interface types.

Right, WASM interface types will "only" handle low-level type conversion for you, but you can embed arbitrary logic within your WASM module....

**Comment by indolering** [comment]
> Before going with the latter choice by default, be sure to consider whether CompCert's license is agreeable.

Which is to say that CompCert doesn't allow for commercial usage and won't work for this project.

**Comment by winitzki** [comment]
Related: https://github.com/dhall-lang/dhall-lang/pull/1440
